Update installation instructions for MacOS (#11957)
PrairieLearn/PrairieLearn: docs/installingNative.md

This page describes the procedure to install and run PrairieLearn fully natively without using Docker. Certain features, such as external graders and workspaces, still require Docker. PrairieLearn supports native execution on macOS, Linux, and Windows inside [WSL 2](https://learn.microsoft.com/en-us/windows/wsl/install).

## Installation

- Install the prerequisites:

  - [Git](https://git-scm.com)
  - [Node.js 20](https://nodejs.org)
  - [Yarn](https://yarnpkg.com)
  - [Python 3.10](https://www.python.org)

>   - [PostgreSQL 17](https://www.postgresql.org)

Reviewer: As I mentioned elsewhere, this is incorrect.

Author: Fixed in 11571eaf4cb629f088c608b49baefe0e835027bd.

    The main prerequisites can be installed with [Homebrew](http://brew.sh/):

    ```sh

>     brew install git graphviz postgresql@17 redis uv d2 node pgvector

Reviewer: Minor nit, as long as we're here.
```suggestion
    brew install git graphviz postgresql@17 redis uv d2 node pgvector

```

Author: Added newline in 91dbf8a.

    # Optional; needed only for some example questions that use LaTeX
    brew install texlive
    ```

    You may want to start up the `postgresql` server on boot, and add binaries to your path:

    ```sh
    brew services start postgresql@17
    # zsh
    echo 'export PATH="$(brew --prefix postgresql@17)/bin:$PATH"' >> ~/.zshrc && . ~/.zshrc
    # bash

>     echo 'export PATH="$(brew --prefix postgresql@17)/bin:$PATH"' >> ~/.bashrc && . ~/.bashrc

Reviewer: This wasn't necessary for me after installing Postgres. Maybe you have conflicting versions? If you run `which psql` in a shell without this, does anything show up? For me, it points to `/opt/homebrew/bin/psql`.

Author: ```
which psql
/opt/homebrew/opt/postgresql@17/bin/psql
```

Author: Removing activation means it doesn't show up for me.

- Make sure the `postgres` database user exists and is a superuser (these might error if the user already exists):

  === "Ubuntu (WSL2)"

      ```sh
      sudo -u postgres psql -c "CREATE USER postgres;"
      sudo -u postgres psql -c "ALTER USER postgres WITH SUPERUSER;"
      sudo -u postgres createdb postgres
      ```

  === "MacOS"

      ```sh
      psql postgres -c "CREATE USER postgres;"
      psql postgres -c "ALTER USER postgres WITH SUPERUSER;"

>       ```

Author: I didn't need to make a database for macOS. Can you suggest a change for the macOS header?

  You may need to restart the PostgreSQL server after changing the file above.

> ## Running

Reviewer: Consider a separate "configuration" heading?

Author: Changed in da99c72.

> If you have your own [PrairieLearn course repository](./requestCourse/index.md), you will need to create the file `PrairieLearn/config.json` with the path of your local course repository. Additionally, if you need support for workspaces, you should provide a path to a directory into which temporary files will be saved when using the in-browser file editor. Here is a sample configuration:

Reviewer: "you should provide a path to a directory into which temporary files will be saved when using the in-browser file editor" is a non sequitur from "if you need support for workspaces".

Author: `Additionally` was removed.

Author: Changed in da99c72.

```json title="config.json"
{
  "courseDirs": ["/Users/mwest/git/pl-tam212", "exampleCourse"],
  "filesRoot": "../filesRoot",
  "workspaceHostHomeDirRoot": "/tmp/workspace",
  "workspaceHomeDirRoot": "/tmp/workspace"
}
```

> ??? note "Advanced configuration"

Reviewer: Review the documentation for admonitions: https://squidfunk.github.io/mkdocs-material/reference/admonitions/. They have different meanings.

>     The full configuration can be found at [`apps/prairielearn/src/lib/config.ts`](https://github.com/PrairieLearn/PrairieLearn/blob/master/apps/prairielearn/src/lib/config.ts).

Reviewer: If this is under a separate heading, I'd be inclined to just link to this outside an admonition.

Author: Changed in [91dbf8a](https://github.com/PrairieLearn/PrairieLearn/pull/11957/commits/91dbf8a49875b10c0d4dd9c47a3f205cd08c1a9d).

- Run the server in development mode to automatically restart when changes are detected:

  ```sh
  make dev
  ```

  !!! warning "macOS"

      If you see the error `"Cannot start s3rver since port 5000 is already in use by process 645."`, toggle off the following setting:

>       `"System Settings..." > General > AirDrop & Handoff > AirPlay Receiver"`

Author: I felt that specifically spelling out the fix was better than the SO link. I can move that additional text into the error message for `start_s3rver.sh` if you would like.

>       `"System Settings..." > General > AirDrop & Handoff > AirPlay Receiver"`

Reviewer: I like the SO link specifically because it includes screenshots, OS-version-specific instructions, and a small bit of context on what it is.

>       `"System Settings..." > General > AirDrop & Handoff > AirPlay Receiver"`

Author: Sure, sounds good.

>       `"System Settings..." > General > AirDrop & Handoff > AirPlay Receiver"`

Author: Changed in [da99c72](https://github.com/PrairieLearn/PrairieLearn/pull/11957/commits/da99c72ac2871a67c4f35e7fd0e5846b3a6db540).

- In a web-browser go to [http://localhost:3000](http://localhost:3000).

- To stop the server, use `Ctrl-C`.

### Production

You can build and run the code to more closely mimic what will happen in production environments:

```sh
make build
make start
```

### Workspaces

- If you need support for [workspaces](workspaces/index.md), ensure Docker is installed and running, and then in a separate terminal run:

```sh
sudo make dev-workspace-host # or sudo make start-workspace-host
```

### Documentation

If you want to preview the documentation, run:

```sh
make preview-docs
```

> ### Testing

Author: Good idea.

> ### Testing

Author: Changed in [da99c72](https://github.com/PrairieLearn/PrairieLearn/pull/11957/commits/da99c72ac2871a67c4f35e7fd0e5846b3a6db540).

> ### Testing

Reviewer: Ah, I hoped that would've been in a new PR so we could merge this one, it was very close. I'll re-review everything.

> ### Testing

Author: Ah, oops. Thanks!

  You may need to restart the PostgreSQL server after changing the file above.

## Configuration

> If you have your own [PrairieLearn course repository](./requestCourse/index.md), you will need to create the file `PrairieLearn/config.json` with the path of your local course repository. If you need support for workspaces, you should provide a path to a directory into which temporary files will be saved when using the in-browser file editor. Here is a sample configuration:

Reviewer: As I mentioned in a previous comment, "you should provide a path to a directory into which temporary files will be saved when using the in-browser file editor" is a non sequitur from "if you need support for workspaces". `filesRoot` is not related to workspace support.

Author: It is unclear to me how to explain `workspaceHostHomeDirRoot`, perhaps you have some  insight there?

Reviewer: Maybe you're getting confused about what "in-browser file editor" is referring to? That's not related to workspaces, though I could see how one might think it is. It's referring to the in-browser file editor for course content itself, like this:

<img width="931" alt="Screenshot 2025-05-12 at 15 47 33" src="https://github.com/user-attachments/assets/1eb010aa-44f0-4610-aaa6-e2c01e917e2c" />

Supporting that is why `filesRoot` must be set.

Author: To clarify -- I had updated the text to:

> If you need support for the in-browser file editor or file uploads, you should set `filesRoot`. If you need support for workspaces, you should provide a path to a directory into which temporary files will be saved. Here is a sample configuration:

I didn't have a great explanation for what exactly `workspaceHostHomeDirRoot` and `workspaceHomeDirRoot` do.

Reviewer: I didn't see that update, there wasn't a commit linked above so I was just responding to the lines that GitHub showed for this comment thread. Your changes look good.

```json title="config.json"
{
  "courseDirs": ["/Users/mwest/git/pl-tam212", "exampleCourse"],
  "filesRoot": "../filesRoot",
  "workspaceHostHomeDirRoot": "/tmp/workspace",
  "workspaceHomeDirRoot": "/tmp/workspace"
}
```

> More information about the `config.json` can be found on the [server configuration](./configJson.md) page.

Reviewer: Nit: in other comments I've been preferring "the XYZ documentation" to "the XYZ page". This is of course just a nit.

Author: Changed in 1539422c261d6f849d3dcc94e7dbaef2d2e39377 and 41f871d284cd02bfd8c98fc64981ce779197b80b.

## Development

> More information on the common commands and actions you do during development can be found on the [quickstart](./quickstart.md) page.

Reviewer: ```suggestion
More information on the development workflow can be found on the [quickstart](./quickstart.md) page.
```

Author: Changed in 11571eaf4cb629f088c608b49baefe0e835027bd.
